lib/plugins/aws/deploy/compile/functions/index.js

```diff
@@ -17,15 +17,42 @@ class AwsCompileFunctions {
     };
   }
 
-  compileRole(role) {
-    if (typeof role === 'object') {
-      // role is an "Fn::GetAtt" object
-      return role;
-    } else if (role.indexOf(':') === -1) {
-      // role is a Logical Role Name
-      return { 'Fn::GetAtt': [role, 'Arn'] };
+  compileRole(newFunction, role) {
+    const compiledFunction = newFunction;
+
+    switch (typeof role) {
+      case 'object':
+        if ('Fn::GetAtt' in role) {
+          // role is an "Fn::GetAtt" object
+          compiledFunction.Properties.Role = role;
+          compiledFunction.DependsOn = [role['Fn::GetAtt'][0]];
+        } else if ('Ref' in role || 'Fn::ImportValue' in role) {
+          // role is a "Ref" or "Fn::ImportValue" object
+          compiledFunction.Properties.Role = role;
+        } else {
+          // TODO: We might want throw a warning here?
+        }
+        break;
+      case 'string':
+        if (role.startsWith('arn:aws')) {
+          // role is a statically definied iam arn
+          compiledFunction.Properties.Role = role;
+        } else if (role === 'IamRoleLambdaExecution') {
+          // role is the default role generated by the framework
+          compiledFunction.Properties.Role = { 'Fn::GetAtt': [role, 'Arn'] };
+          compiledFunction.DependsOn = [
+            'IamPolicyLambdaExecution',
+            'IamRoleLambdaExecution',
+          ];
+        } else {
+          // role is a Logical Role Name
+          compiledFunction.Properties.Role = { 'Fn::GetAtt': [role, 'Arn'] };
+          compiledFunction.DependsOn = [role];
+        }
+        break;
+      default:
+        // TODO: We might want throw a warning here?
```

Comment:
@lcolman Thanks for the tips on regards to `typeof(new Buffer("foo"))`. Will include some checks and tests for it. But TBH not sure how you could inject a `Buffer` from a `serverless.yml`, nonetheless good catch.